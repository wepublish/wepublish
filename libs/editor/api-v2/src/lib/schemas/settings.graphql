# Settings

query SettingsList {
  settingsList {
    id
    name
    value
    settingRestriction {
      maxValue
      minValue
      inputLength
      allowedValues {
        stringChoice
        boolChoice
      }
    }
  }
}

mutation UpdateSetting($name: SettingName!, $value: GraphQLSettingValueType!) {
  updateSetting(name: $name, value: $value) {
    name
    value
  }
}
