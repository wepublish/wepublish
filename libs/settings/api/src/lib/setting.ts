export type Setting<T = unknown> = {
  id: string
  name: SettingName
  value: T | null
  settingRestriction?: SettingRestriction
}

export enum SettingName {
  ALLOW_GUEST_COMMENTING = 'allowGuestCommenting',
  ALLOW_GUEST_COMMENT_RATING = 'allowGuestCommentRating',
  ALLOW_GUEST_POLL_VOTING = 'allowGuestPollVoting',
  SEND_LOGIN_JWT_EXPIRES_MIN = 'sendLoginJwtExpiresMin',
  RESET_PASSWORD_JWT_EXPIRES_MIN = 'resetPasswordJwtExpiresMin',
  PEERING_TIMEOUT_MS = 'peeringTimeoutInMs',
  INVOICE_REMINDER_FREQ = 'invoiceFreqReminder',
  INVOICE_REMINDER_MAX_TRIES = 'invoiceReminderMaxTries',

  MAKE_NEW_SUBSCRIBERS_API_PUBLIC = 'makeNewSubscribersApiPublic',
  MAKE_ACTIVE_SUBSCRIBERS_API_PUBLIC = 'makeActiveSubscribersApiPublic',
  MAKE_RENEWING_SUBSCRIBERS_API_PUBLIC = 'makeRenewingSubscribersApiPublic',
  MAKE_NEW_DEACTIVATIONS_API_PUBLIC = 'makeNewDeactivationsApiPublic',

  MAKE_EXPECTED_REVENUE_API_PUBLIC = 'makeExpectedRevenueApiPublic',
  MAKE_REVENUE_API_PUBLIC = 'makeRevenueApiPublic',
  COMMENT_CHAR_LIMIT = 'commentCharLimit',

  ALLOW_COMMENT_EDITING = 'allowCommentEditing'
}

export type SettingInput<T = unknown> = Pick<Setting<T>, 'value'>

export type CreateSettingArgs<T> = Omit<Setting<T>, 'id'>

export type UpdateSettingArgs<T = unknown> = {
  name: SettingName
  value: T
}

export interface SettingRestriction {
  maxValue?: number
  minValue?: number
  inputLength?: number
  allowedValues?: AllowedSettingVals
}

export type AllowedSettingVals = {
  stringChoice?: string[]
  boolChoice?: boolean
}

export type OptionalSetting = Setting | null
