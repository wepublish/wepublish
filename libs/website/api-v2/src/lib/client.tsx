import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {ComponentType, memo, useMemo} from 'react'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

export const getV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({
    uri: `${apiUrl}/v2`,
    batchMax: 5,
    batchInterval: 20
  })
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV2ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV2ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV2ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV2ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
