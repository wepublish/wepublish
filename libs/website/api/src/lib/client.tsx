import {
  ApolloClient,
  InMemoryCache,
  ApolloProvider,
  ApolloLink,
  NormalizedCacheObject
} from '@apollo/client'
import {BatchHttpLink} from '@apollo/client/link/batch-http'

import {ComponentType, memo, useMemo} from 'react'

export const getV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[],
  cache?: NormalizedCacheObject
) => {
  const httpLink = new BatchHttpLink({uri: `${apiUrl}/v1`, batchMax: 5, batchInterval: 20})
  const link = [...links, httpLink].reduce(
    (links: ApolloLink | undefined, link) => (links ? links.concat(link) : link),
    undefined
  )

  return new ApolloClient({
    link,
    cache: new InMemoryCache().restore(cache ?? {}),
    ssrMode: typeof window === 'undefined'
  })
}

export const useV1ApiClient = (
  apiUrl: string,
  links: ApolloLink[] = [],
  cache?: NormalizedCacheObject
) => useMemo(() => getV1ApiClient(apiUrl, links, cache), [apiUrl, links, cache])

export const createWithV1ApiClient =
  (apiUrl: string, links: ApolloLink[] = [], cache?: NormalizedCacheObject) =>
  <
    // eslint-disable-next-line @typescript-eslint/ban-types
    P extends object
  >(
    ControlledComponent: ComponentType<P>
  ) =>
    memo<P>(props => {
      const client = useV1ApiClient(apiUrl, links, cache)

      return (
        <ApolloProvider client={client}>
          <ControlledComponent {...props} />
        </ApolloProvider>
      )
    })
