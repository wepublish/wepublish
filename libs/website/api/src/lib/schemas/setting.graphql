fragment FullSetting on Setting {
  id
  name
  value
}

query SettingList {
  settings {
    ...FullSetting
  }
}

query Setting($name: String!) {
  setting(name: $name) {
    ...FullSetting
  }
}
