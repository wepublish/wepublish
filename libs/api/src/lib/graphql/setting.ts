import {
  GraphQLObjectType,
  GraphQLNonNull,
  GraphQLString,
  GraphQLInt,
  GraphQLScalarType,
  GraphQLInputObjectType,
  GraphQLID,
  GraphQLList,
  GraphQLEnumType,
  GraphQLBoolean
} from 'graphql'

import {Context} from '../context'
import {Setting, SettingName} from '@wepublish/settings/api'

export const GraphQLSettingValueType = new GraphQLScalarType({
  name: 'Value',
  serialize(Value) {
    return Value
  }
})

export const GraphQLSettingName = new GraphQLEnumType({
  name: 'SettingName',
  values: {
    ALLOW_GUEST_COMMENTING: {value: SettingName.ALLOW_GUEST_COMMENTING},
    ALLOW_GUEST_COMMENT_RATING: {value: SettingName.ALLOW_GUEST_COMMENT_RATING},
    ALLOW_GUEST_POLL_VOTING: {value: SettingName.ALLOW_GUEST_POLL_VOTING},
    SEND_LOGIN_JWT_EXPIRES_MIN: {value: SettingName.SEND_LOGIN_JWT_EXPIRES_MIN},
    RESET_PASSWORD_JWT_EXPIRES_MIN: {value: SettingName.RESET_PASSWORD_JWT_EXPIRES_MIN},
    PEERING_TIMEOUT_MS: {value: SettingName.PEERING_TIMEOUT_MS},
    INVOICE_REMINDER_FREQ: {value: SettingName.INVOICE_REMINDER_FREQ},
    INVOICE_REMINDER_MAX_TRIES: {value: SettingName.INVOICE_REMINDER_MAX_TRIES},
    MAKE_ACTIVE_SUBSCRIBERS_API_PUBLIC: {value: SettingName.MAKE_ACTIVE_SUBSCRIBERS_API_PUBLIC},
    MAKE_NEW_SUBSCRIBERS_API_PUBLIC: {value: SettingName.MAKE_NEW_SUBSCRIBERS_API_PUBLIC},
    MAKE_RENEWING_SUBSCRIBERS_API_PUBLIC: {value: SettingName.MAKE_RENEWING_SUBSCRIBERS_API_PUBLIC},
    MAKE_NEW_DEACTIVATIONS_API_PUBLIC: {value: SettingName.MAKE_NEW_DEACTIVATIONS_API_PUBLIC},
    MAKE_EXPECTED_REVENUE_API_PUBLIC: {value: SettingName.MAKE_EXPECTED_REVENUE_API_PUBLIC},
    MAKE_REVENUE_API_PUBLIC: {value: SettingName.MAKE_REVENUE_API_PUBLIC},
    COMMENT_CHAR_LIMIT: {value: SettingName.COMMENT_CHAR_LIMIT},
    ALLOW_COMMENT_EDITING: {value: SettingName.ALLOW_COMMENT_EDITING}
  }
})

export const GraphQLAllowedSettingVals = new GraphQLObjectType({
  name: 'AllowedSettingVals',
  fields: {
    stringChoice: {type: GraphQLList(GraphQLString)},
    boolChoice: {type: GraphQLBoolean}
  }
})

export const GraphQLSettingRestriction = new GraphQLObjectType({
  name: 'SettingRestriction',
  fields: {
    maxValue: {type: GraphQLInt},
    minValue: {type: GraphQLInt},
    inputLength: {type: GraphQLInt},
    allowedValues: {type: GraphQLAllowedSettingVals}
  }
})

export const GraphQLSettingRestrictionInput = new GraphQLInputObjectType({
  name: 'SettingRestrictionInput',
  fields: {
    maxValue: {type: GraphQLInt},
    minValue: {type: GraphQLInt},
    inputLength: {type: GraphQLInt},
    allowedValues: {type: GraphQLList(GraphQLAllowedSettingVals)}
  }
})

export const GraphQLSettingInput = new GraphQLInputObjectType({
  name: 'SettingInput',
  fields: {
    value: {type: GraphQLSettingValueType}
  }
})

export const GraphQLUpdateSettingArgs = new GraphQLInputObjectType({
  name: 'UpdateSettingArgs',
  fields: {
    name: {type: GraphQLNonNull(GraphQLSettingName)},
    value: {type: GraphQLNonNull(GraphQLSettingValueType)}
  }
})

export const GraphQLSettingsInput = new GraphQLInputObjectType({
  name: 'SettingsInput',
  fields: {
    value: {type: GraphQLList(GraphQLUpdateSettingArgs)}
  }
})

export const GraphQLSetting = new GraphQLObjectType<Setting, Context>({
  name: 'Setting',
  fields: {
    id: {type: GraphQLNonNull(GraphQLID)},
    name: {type: GraphQLNonNull(GraphQLSettingName)},
    value: {type: GraphQLNonNull(GraphQLSettingValueType)},
    settingRestriction: {type: GraphQLSettingRestriction}
  }
})
