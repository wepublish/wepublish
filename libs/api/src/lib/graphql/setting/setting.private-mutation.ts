import {Prisma, PrismaClient} from '@prisma/client'
import {CanUpdateSettings} from '@wepublish/permissions/api'
import {Context} from '../../context'
import {SettingRestriction, UpdateSettingArgs} from '@wepublish/settings/api'
import {NotFound} from '../../error'
import {checkSettingRestrictions} from '../../utility'
import {authorise} from '../permissions'

export const updateSettings = async (
  value: UpdateSettingArgs[],
  authenticate: Context['authenticate'],
  prisma: PrismaClient
) => {
  const {roles} = authenticate()
  authorise(CanUpdateSettings, roles)

  for (const {name, value: newVal} of value) {
    const fullSetting = await prisma.setting.findUnique({
      where: {name}
    })

    if (!fullSetting) {
      throw new NotFound('setting', name)
    }

    const currentVal = fullSetting.value
    const restriction = fullSetting.settingRestriction
    checkSettingRestrictions(newVal, currentVal, restriction as SettingRestriction)
  }

  return prisma.$transaction(
    (value as UpdateSettingArgs[]).map(({name, value: val}) =>
      prisma.setting.update({
        where: {
          name
        },
        data: {
          value: val as Prisma.InputJsonValue
        }
      })
    )
  )
}
