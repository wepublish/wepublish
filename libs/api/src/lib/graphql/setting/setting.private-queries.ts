import {PrismaClient} from '@prisma/client'
import {CanGetSettings} from '@wepublish/permissions/api'
import {Context} from '../../context'
import {authorise} from '../permissions'

export const getSetting = (
  name: string,
  authenticate: Context['authenticate'],
  setting: PrismaClient['setting']
) => {
  const {roles} = authenticate()
  authorise(CanGetSettings, roles)

  return setting.findUnique({
    where: {name}
  })
}

export const getSettings = (
  authenticate: Context['authenticate'],
  setting: PrismaClient['setting']
) => {
  const {roles} = authenticate()
  authorise(CanGetSettings, roles)

  return setting.findMany({})
}
