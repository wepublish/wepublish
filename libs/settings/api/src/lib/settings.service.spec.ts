import {Test, TestingModule} from '@nestjs/testing'
import {SettingsService} from './settings.service'
import {PrismaModule} from '@wepublish/nest-modules'
import {PrismaClient, Setting} from '@prisma/client'
import {SettingName} from './setting'
import {GraphQLSettingValueType} from './settings.model'

describe('SettingsService', () => {
  let service: SettingsService
  let prisma: PrismaClient

  beforeAll(async () => {
    const module: TestingModule = await Test.createTestingModule({
      imports: [PrismaModule],
      providers: [SettingsService, GraphQLSettingValueType]
    }).compile()

    prisma = module.get<PrismaClient>(PrismaClient)
    service = module.get<SettingsService>(SettingsService)
  })

  test('should be defined', () => {
    expect(service).toBeDefined()
  })

  test('should list settings', async () => {
    const mockSettings: Setting[] = [
      {
        id: '1',
        name: 'setting1',
        value: 'value1',
        createdAt: new Date('1/1/2020'),
        modifiedAt: new Date('2/1/2020'),
        settingRestriction: null
      },
      {
        id: '2',
        name: 'setting2',
        value: 'value2',
        createdAt: new Date('1/1/2020'),
        modifiedAt: new Date('2/1/2020'),
        settingRestriction: null
      }
    ]

    const mockFunction = jest.spyOn(prisma.setting, 'findMany').mockResolvedValue(mockSettings)

    const result = await service.settingsList()
    expect(result).toMatchSnapshot()
    expect(mockFunction.mock.calls[0][0]).toMatchSnapshot()
  })

  test('should fetch a single setting by ID', async () => {
    const mockSetting: Setting = {
      id: '1',
      name: 'setting1',
      value: 'value1',
      createdAt: new Date('1/1/2020'),
      modifiedAt: new Date('2/1/2020'),
      settingRestriction: null
    }

    const mockFunction = jest.spyOn(prisma.setting, 'findUnique').mockResolvedValue(mockSetting)

    const id = '1'
    const result = await service.setting(id)
    expect(result).toMatchSnapshot()
    expect(mockFunction.mock.calls[0][0]).toMatchSnapshot()
  })

  test('should update settings', async () => {
    const updateInput = {name: SettingName.ALLOW_COMMENT_EDITING, value: false}

    const updatedSetting: Setting = {
      id: '1',
      name: SettingName.ALLOW_COMMENT_EDITING,
      value: false,
      createdAt: new Date('1/1/2020'),
      modifiedAt: new Date('2/1/2020'),
      settingRestriction: {
        allowedValues: {
          boolChoice: true
        }
      }
    }

    jest.spyOn(prisma.setting, 'findUnique').mockResolvedValue(updatedSetting)
    jest.spyOn(prisma.setting, 'update').mockResolvedValue(updatedSetting)

    const result = await service.updateSetting(updateInput)
    expect(result).toMatchSnapshot({
      modifiedAt: expect.any(Date),
      createdAt: expect.any(Date),
      id: expect.any(String)
    })
  })
})
