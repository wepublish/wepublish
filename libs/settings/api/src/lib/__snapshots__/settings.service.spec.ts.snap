// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsService should fetch a single setting by ID 1`] = `
{
  "createdAt": 2020-01-01T00:00:00.000Z,
  "id": "1",
  "modifiedAt": 2020-02-01T00:00:00.000Z,
  "name": "setting1",
  "settingRestriction": null,
  "value": "value1",
}
`;

exports[`SettingsService should fetch a single setting by ID 2`] = `
{
  "where": {
    "id": "1",
  },
}
`;

exports[`SettingsService should list settings 1`] = `
[
  {
    "createdAt": 2020-01-01T00:00:00.000Z,
    "id": "1",
    "modifiedAt": 2020-02-01T00:00:00.000Z,
    "name": "setting1",
    "settingRestriction": null,
    "value": "value1",
  },
  {
    "createdAt": 2020-01-01T00:00:00.000Z,
    "id": "2",
    "modifiedAt": 2020-02-01T00:00:00.000Z,
    "name": "setting2",
    "settingRestriction": null,
    "value": "value2",
  },
]
`;

exports[`SettingsService should list settings 2`] = `
{
  "orderBy": {
    "createdAt": "desc",
  },
  "where": {},
}
`;

exports[`SettingsService should update settings 1`] = `
{
  "createdAt": Any<Date>,
  "id": Any<String>,
  "modifiedAt": Any<Date>,
  "name": "allowCommentEditing",
  "settingRestriction": {
    "allowedValues": {
      "boolChoice": true,
    },
  },
  "value": false,
}
`;
